src/ecmascript/spidermonkey/localstorage.c

```diff
@@ -0,0 +1,208 @@
+#include "util/memory.h"
+#include "util/string.h"
+#include "viewer/text/draw.h"
+#include "viewer/text/form.h"
+#include "viewer/text/link.h"
+#include "viewer/text/vs.h"
+
+#include <time.h>
+#include "document/renderer.h"
+#include "document/refresh.h"
+#include "terminal/screen.h"
+
+/* IMPLEMENTS READ FROM STORAGE USING SQLITE DATABASE */
+static unsigned char *
+readFromStorage(unsigned char *key)
+{
+
+	char * val = "";
+	val = (unsigned char *) malloc (32384);
+
+	if (local_storage_ready==0)
+	{
+		db_prepare_structure(local_storage_filename);
+		local_storage_ready=1;
+	}
+
+	val = db_query_by_key(local_storage_filename, key);
+
+	// DBG(log,"Read: %s %s %s",local_storage_filename, key, val);
+
+	return (val);
+}
+
+/* IMPLEMENTS SAVE TO STORAGE USING SQLITE DATABASE */
+static void
+saveToStorage(unsigned char *key, unsigned char *val)
+{
+
+	if (local_storage_ready==0) {
+		db_prepare_structure(local_storage_filename);
+		local_storage_ready=1;
+	}
+
+	int rows_affected=0;
+
+	rows_affected=db_update_set(local_storage_filename, key, val);
+
+	if (rows_affected==0) {
+		rows_affected=db_insert_into(local_storage_filename, key, val);
+	}
+
+	// DBG(log, "UPD ROWS: %d KEY: %s VAL: %s",rows_affected,key,val);
+
+}
+
+static bool localstorage_get_property(JSContext *ctx, JS::HandleObject hobj, JS::HandleId hid, JS::MutableHandleValue hvp);
+
+JSClassOps localstorage_ops = {
+	JS_PropertyStub, nullptr,
+	localstorage_get_property, JS_StrictPropertyStub,
+	nullptr, nullptr, nullptr
+};
+
+/* Each @localstorage_class object must have a @window_class parent.  */
+const JSClass localstorage_class = {
+	"localStorage",
+	JSCLASS_HAS_PRIVATE,
+	&localstorage_ops
+};
+
+const JSPropertySpec localstorage_props[] = {
+	{ NULL }
+};
+
+///* @localstorage_class.getProperty */
+static bool
+localstorage_get_property(JSContext *ctx, JS::HandleObject hobj, JS::HandleId hid, JS::MutableHandleValue hvp)
+{
+	JSObject *parent_win;	/* instance of @window_class */
+
+	return(true);
+}
+
+static bool localstorage_setitem(JSContext *ctx, unsigned int argc, JS::Value *vp);
+static bool localstorage_getitem(JSContext *ctx, unsigned int argc, JS::Value *vp);
+
+const spidermonkeyFunctionSpec localstorage_funcs[] = {
+	{ "setItem",		localstorage_setitem,	2 },
+	{ "getItem",		localstorage_getitem,	1 },
+	{ NULL }
+};
+
+static bool
+localstorage_getitem(JSContext *ctx, unsigned int argc, JS::Value *vp)
+{
+	//jsval val;
+	JSCompartment *comp = js::GetContextCompartment(ctx);
+
+	if (!comp)
+	{
+		return false;
+	}
+
+	struct ecmascript_interpreter *interpreter = JS_GetCompartmentPrivate(comp);
+	JS::CallArgs args = CallArgsFromVp(argc, vp);
+        unsigned char *key = JS_EncodeString(ctx, args[0].toString());
+	//DBG("localstorage get by key: %s\n", args);
+
+	if (argc != 1)
+	{
+	       args.rval().setBoolean(false);
+	       return(true);
+	}
+
+	unsigned char *val;
+        val = (unsigned char * ) malloc(32000);
+        val = readFromStorage(key);
```

Comment:
Do not malloc here. Memory will be allocated by readFromStorage.

Reply:
O.K. agreed, Fixed.